File: py/graph/eulerian_path/undirected/Hierholzer.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class Hierholzer:
    def search(self, graphs: List[VertexNode]) -> List[int]:
        start = self.__getStart(graphs)
        if not ~start:
            return []

        visited, result = [[False for _ in range(len(graphs))] for _ in range(len(graphs))], []
        self.__search(graphs, start, visited, result)
        result.reverse()
        return result

    def __search(self, graphs: List[VertexNode], start: int, visited: List[List[bool]], result: List[List[int]]) -> None:
        edge = graphs[start].next
        while edge:
            edgeIndex = edge.index
            if not visited[start][edgeIndex]:
                visited[start][edgeIndex], visited[edgeIndex][start] = True, True
                self.__search(graphs, edgeIndex, visited, result)
            edge = edge.next
        result.append(start)

    def __getStart(self, graphs: List[VertexNode]) -> int:
        start, oddCount = 0, 0
        for i in range(len(graphs) - 1, -1, -1):
            count, edge = 0, graphs[i].next
            while edge:
                count += 1
                edge = edge.next
            if count & 1:
                start = i
                oddCount += 1
                if oddCount > 2:
                    break
        return start if not oddCount or 2 == oddCount else -1

File: cpp/graph/eulerian_path/undirected/Hierholzer.cpp
#include <iostream>
#include <vector>
using namespace std;

struct EdgeNode {
    int index;
    int weight;
    EdgeNode *next;

    EdgeNode(int index, int weight) : index(index), weight(weight), next(nullptr) {}
};

struct VertexNode {
    int value;
    EdgeNode *next;

    VertexNode(int value) : value(value), next(nullptr) {}
};

class Hierholzer {
public:
    vector<int> search(vector<VertexNode *> &graphs) {
        int start = getStart(graphs);
        if (!~start) {
            return vector<int>(0);
        }

        vector<vector<bool>> visited(graphs.size(), vector<bool>(graphs.size(), false));
        vector<int> result(0);
        search(graphs, start, visited, result);
        reverse(result.begin(), result.end());
        return result;
    }

private:
    void search(vector<VertexNode *> &graphs, int start, vector<vector<bool>> &visited, vector<int> &result) {
        EdgeNode *edge = graphs[start]->next;
        while (edge) {
            int edgeIndex = edge->index;
            if (!visited[start][edgeIndex]) {
                visited[start][edgeIndex] = true;
                visited[edgeIndex][start] = true;
                search(graphs, edgeIndex, visited, result);
            }
            edge = edge->next;
        }
        result.emplace_back(start);
    }

    int getStart(vector<VertexNode *> &graphs) {
        int start = 0;
        int oddCount = 0;
        for (int i = graphs.size() - 1; i >= 0; --i) {
            int count = 0;
            EdgeNode *edge = graphs[i]->next;
            while (edge) {
                ++count;
                edge = edge->next;
            }
            if (count & 1) {
                start = i;
                ++oddCount;
            }
        }
        return (!oddCount || 2 == oddCount) ? start : -1;
    }
};
